mozetl/engagement/retention/job.py

```diff
@@ -0,0 +1,166 @@
+from schema import retention_schema
+
+
+def valid_pcd(pcd, client_date):
+    pcd_format = F.date_format(pcd, 'yyyy-MM-dd')
+    is_valid_pcd = [
+        pcd_format.isNotNull(),
+        (pcd_format > DEFAULT_DATE),
+        (pcd <= client_date),
+    ]
+    return (
+        F.when(reduce(operator.__and__, is_valid_pcd), pcd)
+        .otherwise(F.lit(None))
+    )
+
+
+def transform(main_summary, start_ds):
+
+    client_date = utils.to_datetime('subsession_start_date', "yyyy-MM-dd")
+
+    pcd = valid_pcd(
+        F.from_unixtime(F.col("profile_creation_date") * SECONDS_PER_DAY),
+        client_date
+    )
+
+    days_since_creation = F.datediff(client_date, pcd)
+
+    # Bug 1289573: Support values like "mozilla86" and "mozilla86-utility-existing"
+    is_funnelcake = F.col('distribution_id').rlike("^mozilla[0-9]+.*$")
+
+    subsession_length = (
+        F.when(F.col('subsession_length') > MAX_SUBSESSION_LENGTH, MAX_SUBSESSION_LENGTH)
+        .otherwise(F.when(F.col('subsession_length') < 0, 0)
+                   .otherwise(F.col('subsession_length')))
+    )
+    usage_hours = subsession_length / SECONDS_PER_HOUR
+
+    mapping = {
+        # attributes
+        'client_id': None,
+        'subsession_start': F.date_format(client_date, "yyyy-MM-dd"),
+        'profile_creation': F.date_format(pcd, "yyyy-MM-dd"),
+        'days_since_creation': (
+            F.when(days_since_creation < 0, F.lit(-1))  # -1 is the value for bad data
+            .otherwise(days_since_creation)
+            .cast("long")),
+        'channel': 'normalized_channel',
+        'app_version': None,
+        'geo': in_top_countries("country"),
+        'distribution_id': None,
+        'is_funnelcake': is_funnelcake,
+        'source': 'attribution.source',
+        'medium': 'attribution.medium',
+        'campaign': 'attribution.campaign',
+        'content': 'attribution.content',
+        'sync_usage': sync_usage(
+            "sync_count_desktop",
+            "sync_count_mobile",
+            "sync_configured"
+        ),
+        'is_active': client_date <= utils.to_datetime(F.lit(start_ds)),
+        # metrics
+        'usage_hours': usage_hours,
+        'sum_squared_usage_hours': usage_hours ** 2,
+        'total_uri_count':
+            F.col('scalar_parent_browser_engagement_total_uri_count').cast('long'),
+        'unique_domains_count':
+            F.col('scalar_parent_browser_engagement_unique_domains_count').cast('long'),
+    }
+    expr = utils.build_col_expr(mapping)
+
+    cleaned_data = (
+        main_summary
+        .select(expr)
+        .fillna({
+            'profile_creation': DEFAULT_DATE,
+            "days_since_creation": -1,
+        })
+        .fillna(0)
+        .fillna(0.0)
+        .fillna('unknown')
+        .select([field.name for field in retention_schema.fields])
+    )
+
+    return cleaned_data
+
+
+def save(cleaned, path):
+    (
+        cleaned
+        .write
+        .parquet(path, mode="overwrite")
+    )
```

Comment:
What happened here? This can be:
```python
cleaned.write.parquet(path, mode="overwrite")
```

Reply:
When you realize parentheses can be used to format chained method calls, everything looks like a nail.